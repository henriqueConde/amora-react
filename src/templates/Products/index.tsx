import Base from 'templates/Base'
import { KeyboardArrowDown as ArrowDown } from '@styled-icons/material-outlined/KeyboardArrowDown'

import ExploreSidebar, { ItemProps } from 'components/ExploreSidebar'
import ProductCard, { ProductCardProps } from 'components/ProductCard'
import { Grid } from 'components/Grid'

import * as S from './styles'
import { GET_PRODUCTS } from 'graphql/generated/GET_PRODUCTS'
import { mapProducts } from 'utils/queriesMappers'
import { useEffect, useState } from 'react'
import { useQueryProducts } from 'hooks/useQueryProducts'
import Loading from 'components/Loading'
import {
  parseQueryStringToFilter,
  parseQueryStringToWhere
} from 'utils/filters'
import { useRouter } from 'next/router'
import { ParsedUrlQueryInput } from 'querystring'

export type ProductsTemplateProps = {
  products?: ProductCardProps[]
  filterItems: ItemProps[]
}

const ProductsTemplate = ({ filterItems }: ProductsTemplateProps) => {
  const { push, query } = useRouter()
  console.log(query)
  const { data, fetchMore, error, loading } = useQueryProducts({
    notifyOnNetworkStatusChange: true,
    variables: {
      limit: 2,
      where: query
        ? parseQueryStringToWhere({ queryString: query, filterItems })
        : null,
      sort: query.sort ? [query?.sort as string] : null
    }
  })

  const [availableProds, setAvailableProds] = useState<ProductCardProps[]>()
  const [page, setPage] = useState(1)
  const [isLastPage, setIsLastPage] = useState(false)

  const handleFilter = (items: ParsedUrlQueryInput) => {
    setIsLastPage(false)
    push({
      pathname: '/products',
      query: items
    })
    return
  }

  const handleShowMore = () => {
    const limit = 3
    if (availableProds) {
      fetchMore({
        variables: { limit, start: page },
        updateQuery: (preResult, { fetchMoreResult }) => {
          const fetchMoreProds = mapProducts(fetchMoreResult)
          const newAvailableProds = [...availableProds, ...fetchMoreProds]
          if (fetchMoreProds.length < limit) {
            setIsLastPage(true)
          }
          setAvailableProds(newAvailableProds)
          return preResult
        }
      })
    }
  }

  useEffect(() => {
    data && setAvailableProds(mapProducts(data as GET_PRODUCTS))
  }, [data])

  useEffect(() => {
    availableProds && setPage(availableProds.length)
  }, [availableProds])

  return (
    <Base>
      <S.Main>
        {error && <p>{error.message}</p>}
        <>
          <ExploreSidebar
            initialValues={parseQueryStringToFilter({
              queryString: query,
              filterItems
            })}
            items={filterItems}
            onFilter={handleFilter}
          />

          <section>
            <Grid>
              {(availableProds as ProductCardProps[])?.map((item) => (
                <ProductCard key={item.title} {...item} />
              ))}
            </Grid>
            {loading ? (
              <Loading />
            ) : (
              !isLastPage && (
                <S.ShowMore role="button" onClick={handleShowMore}>
                  <p>Show More</p>
                  <ArrowDown size={35} />
                </S.ShowMore>
              )
            )}
          </section>
        </>
      </S.Main>
    </Base>
  )
}

export default ProductsTemplate
