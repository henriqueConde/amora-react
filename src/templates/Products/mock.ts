export const productsResponseMock = {
  data: {
    products: {
      data: [
        {
          attributes: {
            name: 'Coffee',
            slug: 'coffee',
            price: 164.9,
            brand: {
              data: {
                attributes: { name: 'Lumiere Eau', __typename: 'Brand' },
                __typename: 'BrandEntity'
              },
              __typename: 'BrandEntityResponse'
            },
            cover: {
              data: {
                attributes: {
                  url: '/uploads/ma_cherie_cover_b5ac84d5f2.png',
                  __typename: 'UploadFile'
                },
                __typename: 'UploadFileEntity'
              },
              __typename: 'UploadFileEntityResponse'
            },
            __typename: 'Product'
          },
          __typename: 'ProductEntity'
        },
        {
          attributes: {
            name: 'Ma Chérie',
            slug: 'ma-cherie',
            price: 95.9,
            brand: { data: null, __typename: 'BrandEntityResponse' },
            cover: {
              data: {
                attributes: {
                  url: '/uploads/ma_cherie_cover_b5ac84d5f2.png',
                  __typename: 'UploadFile'
                },
                __typename: 'UploadFileEntity'
              },
              __typename: 'UploadFileEntityResponse'
            },
            __typename: 'Product'
          },
          __typename: 'ProductEntity'
        },
        {
          attributes: {
            name: 'Glamour',
            slug: 'glamour',
            price: 123.9,
            brand: { data: null, __typename: 'BrandEntityResponse' },
            cover: {
              data: {
                attributes: {
                  url: '/uploads/glamour_cover_5d3e1fa760.png',
                  __typename: 'UploadFile'
                },
                __typename: 'UploadFileEntity'
              },
              __typename: 'UploadFileEntityResponse'
            },
            __typename: 'Product'
          },
          __typename: 'ProductEntity'
        },
        {
          attributes: {
            name: 'Florata',
            slug: 'florata',
            price: 103,
            brand: {
              data: {
                attributes: { name: 'Elyseé Parfum', __typename: 'Brand' },
                __typename: 'BrandEntity'
              },
              __typename: 'BrandEntityResponse'
            },
            cover: {
              data: {
                attributes: {
                  url: '/uploads/florata_cover_fa34c70658.png',
                  __typename: 'UploadFile'
                },
                __typename: 'UploadFileEntity'
              },
              __typename: 'UploadFileEntityResponse'
            },
            __typename: 'Product'
          },
          __typename: 'ProductEntity'
        },
        {
          attributes: {
            name: 'Lily',
            slug: 'lily',
            price: 233.9,
            brand: { data: null, __typename: 'BrandEntityResponse' },
            cover: {
              data: {
                attributes: {
                  url: '/uploads/glamour_cover_5d3e1fa760.png',
                  __typename: 'UploadFile'
                },
                __typename: 'UploadFileEntity'
              },
              __typename: 'UploadFileEntityResponse'
            },
            __typename: 'Product'
          },
          __typename: 'ProductEntity'
        }
      ],
      __typename: 'ProductEntityResponseCollection'
    }
  }
}
