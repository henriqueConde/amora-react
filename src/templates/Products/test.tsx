import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/tests/helpers'
import perfumesMock from 'components/ProductCardSlider/mock'
import filterItemsMock from 'components/ExploreSidebar/mock'

import Products from '.'

jest.mock('templates/Base', () => ({
  __esModule: true,
  default: function Mock({ children }: { children: React.ReactNode }) {
    return <div data-testid="Mock Base">{children}</div>
  }
}))

jest.mock('components/ExploreSidebar', () => ({
  __esModule: true,
  default: function Mock({ children }: { children: React.ReactNode }) {
    return <div data-testid="Mock ExploreSidebar">{children}</div>
  }
}))

jest.mock('components/ProductCard', () => ({
  __esModule: true,
  default: function Mock() {
    return <div data-testid="Mock ProductCard" />
  }
}))

describe('<Products />', () => {
  it('should render sections', () => {
    renderWithTheme(
      <Products filterItems={filterItemsMock} products={[perfumesMock[0]]} />
    )

    expect(screen.getByTestId('Mock ExploreSidebar')).toBeInTheDocument()
    expect(screen.getByTestId('Mock ProductCard')).toBeInTheDocument()

    expect(
      screen.getByRole('button', { name: /show more/i })
    ).toBeInTheDocument()
  })
})
