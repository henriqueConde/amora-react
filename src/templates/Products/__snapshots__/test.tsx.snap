// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Products /> should render sections 1`] = `
.c4 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c2 {
  display: grid;
  grid-template-columns: repeat(auto-fill,minmax(25rem,1fr));
  grid-gap: 3.2rem;
  margin: 3.2rem 0;
}

.c0 {
  width: 100%;
  max-width: 130rem;
  margin-left: auto;
  margin-right: auto;
  padding-left: calc(3.2rem / 2);
  padding-right: calc(3.2rem / 2);
}

.c3 {
  color: #FAFAFA;
  text-align: center;
  padding: 3.2rem;
  text-transform: uppercase;
  font-weight: bold;
  cursor: pointer;
}

.c3 > svg {
  color: #F231A5;
}

@media (min-width:768px) {
  .c1 {
    display: grid;
    grid-template-columns: 26rem 1fr;
    gap: 3.2rem;
  }
}

<div>
  <div
    data-testid="Mock Base"
  >
    <div
      class="c0 c1"
    >
      <div
        data-testid="Mock ExploreSidebar"
      />
      <section>
        <div
          class="c2"
        />
        <div
          class="c3"
          role="button"
        >
          <p>
            Show More
          </p>
          <svg
            aria-hidden="true"
            class="c4"
            fill="currentColor"
            focusable="false"
            height="35"
            viewBox="0 0 24 24"
            width="35"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
            />
          </svg>
        </div>
      </section>
    </div>
  </div>
</div>
`
